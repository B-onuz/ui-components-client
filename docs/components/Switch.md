# Switch (Component)

Componente Switch, criado com base no checkbox nativo.

## Como utilizar

```js
import Switch from 'ui-components/lib/components/Switch'
```

```jsx
import Switch from 'ui-components/lib/components/Switch'
import React, { useState } from 'react'

const SwitchExample = () => {
  const [checked, setChecked] = useState(false)
  const handleChangeSwitch = (e) => {
    setChecked(e.target.checked)
  }
  return <Switch checked={checked} onChange={handleChangeSwitch} />
}
```

## Variações

### Color

```jsx
<Switch color={'primary'} />
```

Escala no tema: `theme.colors`

## Styled-system props

- space https://styled-system.com/table#space
