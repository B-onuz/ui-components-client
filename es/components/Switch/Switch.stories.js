'use strict'

Object.defineProperty(exports, '__esModule', {
  value: true,
})
exports.SwitchExample = exports['default'] = void 0

var _react = _interopRequireDefault(require('react'))

var _Switch = _interopRequireDefault(require('./Switch'))

var _this = void 0,
  _jsxFileName = '/home/junior/Fontes/ui-components-client/src/components/Switch/Switch.stories.js'

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj }
}

var _default = {
  title: 'Switch',
  component: _Switch['default'],
}
exports['default'] = _default

var SwitchExample = function SwitchExample() {
  return /*#__PURE__*/ _react['default'].createElement(
    _react['default'].Fragment,
    null,
    /*#__PURE__*/ _react['default'].createElement(_Switch['default'], {
      m: 10,
      color: 'primary',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 11,
        columnNumber: 5,
      },
    }),
    /*#__PURE__*/ _react['default'].createElement(_Switch['default'], {
      m: 10,
      color: 'secondary',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 12,
        columnNumber: 5,
      },
    }),
    /*#__PURE__*/ _react['default'].createElement(_Switch['default'], {
      m: 10,
      color: 'olive',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 13,
        columnNumber: 5,
      },
    }),
    /*#__PURE__*/ _react['default'].createElement(_Switch['default'], {
      m: 10,
      color: 'blue',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 14,
        columnNumber: 5,
      },
    }),
    /*#__PURE__*/ _react['default'].createElement(_Switch['default'], {
      m: 10,
      color: 'green',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 15,
        columnNumber: 5,
      },
    }),
    /*#__PURE__*/ _react['default'].createElement(_Switch['default'], {
      m: 10,
      color: 'yellow',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 16,
        columnNumber: 5,
      },
    }),
    /*#__PURE__*/ _react['default'].createElement(_Switch['default'], {
      m: 10,
      color: 'danger',
      __self: _this,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 17,
        columnNumber: 5,
      },
    })
  )
}

exports.SwitchExample = SwitchExample
