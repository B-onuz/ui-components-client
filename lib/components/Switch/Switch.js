"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

var _this = void 0,
    _jsxFileName = "/Users/master/projects/bonuz/ui-components/src/components/Switch/Switch.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject3() {
  var data = _taggedTemplateLiteralLoose(["\n  border-color: rgb(223, 223, 223);\n  box-shadow: rgb(223, 223, 223) 1px 1px 3px 0px inset;\n  transition: border 0.4s, box-shadow 0.4s;\n  background-color: #f7f7f7;\n  border: 1px solid #dfdfdf;\n  border-radius: 20px;\n  cursor: pointer;\n  display: inline-block;\n  height: 20px;\n  position: relative;\n  vertical-align: middle;\n  width: 40px;\n  user-select: none;\n  overflow: hidden;\n  box-sizing: content-box;\n  background-clip: content-box;\n  -webkit-mask-image: -webkit-radial-gradient(white, black);\n  ", "\n  ", "\n  input {\n    display: none;\n    & + ", " {\n      transform: translate(0);\n    }\n\n    &:checked + ", " {\n      transform: translate(calc(50% + 10px));\n    }\n\n    &:checked ~ ", " {\n      background: ", ";\n      opacity: 1;\n      transform: translate(0);\n    }\n  }\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  position: absolute;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0;\n  z-index: 1;\n  transform: translate(calc(-55% - 25px));\n  transition: 0.25s ease-out;\n  opacity: 0;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  display: block;\n  background: #fff;\n  position: relative;\n  border-radius: 100%;\n  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.4);\n  height: 20px;\n  width: 20px;\n  top: 0;\n  display: block;\n  transform: translate(25px);\n  transition: 0.2s;\n  z-index: 2;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var SwitchToggle = _styledComponents["default"].div(_templateObject());

var SwitchBg = _styledComponents["default"].div(_templateObject2());

var Wrapper = _styledComponents["default"].label(_templateObject3(), _styledSystem.space, function (props) {
  if (props.loading) {
    return "\n          opacity: .5;\n      ";
  }
}, SwitchToggle, SwitchToggle, SwitchBg, function (_ref) {
  var theme = _ref.theme,
      color = _ref.color;
  return theme.colors[color];
});

var Switch = function Switch(props) {
  var onChange = props.onChange,
      checked = props.checked,
      id = props.id,
      _props$loading = props.loading,
      loading = _props$loading === void 0 ? false : _props$loading,
      _props$color = props.color,
      color = _props$color === void 0 ? 'primary' : _props$color,
      rest = _objectWithoutPropertiesLoose(props, ["onChange", "checked", "id", "loading", "color"]);

  return /*#__PURE__*/_react["default"].createElement(Wrapper, _extends({
    htmlFor: id,
    loading: loading,
    color: color
  }, rest, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 79,
      columnNumber: 5
    }
  }), /*#__PURE__*/_react["default"].createElement("input", _extends({
    type: "checkbox",
    id: id,
    checked: checked,
    onChange: onChange
  }, rest, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 80,
      columnNumber: 7
    }
  })), /*#__PURE__*/_react["default"].createElement(SwitchToggle, {
    color: color,
    loading: loading,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 81,
      columnNumber: 7
    }
  }), /*#__PURE__*/_react["default"].createElement(SwitchBg, {
    color: color,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 82,
      columnNumber: 7
    }
  }));
};

var _default = Switch;
exports["default"] = _default;
module.exports = exports.default;