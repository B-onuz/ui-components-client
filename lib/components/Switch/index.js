"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _Switch = _interopRequireDefault(require("./Switch"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = _Switch["default"];
exports["default"] = _default;
module.exports = exports.default;