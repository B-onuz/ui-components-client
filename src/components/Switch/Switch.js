import React from 'react'
import styled from 'styled-components'
import { space } from 'styled-system'

const SwitchToggle = styled.div`
  display: block;
  background: #fff;
  position: relative;
  border-radius: 100%;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.4);
  height: 20px;
  width: 20px;
  top: 0;
  display: block;
  transform: translate(25px);
  transition: 0.2s;
  z-index: 2;
`

const SwitchBg = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  transform: translate(calc(-55% - 25px));
  transition: 0.25s ease-out;
  opacity: 0;
`

const Wrapper = styled.label`
  border-color: rgb(223, 223, 223);
  box-shadow: rgb(223, 223, 223) 1px 1px 3px 0px inset;
  transition: border 0.4s, box-shadow 0.4s;
  background-color: #f7f7f7;
  border: 1px solid #dfdfdf;
  border-radius: 20px;
  cursor: pointer;
  display: inline-block;
  height: 20px;
  position: relative;
  vertical-align: middle;
  width: 40px;
  user-select: none;
  overflow: hidden;
  box-sizing: content-box;
  background-clip: content-box;
  -webkit-mask-image: -webkit-radial-gradient(white, black);
  ${space}
  ${(props) => {
    if (props.loading) {
      return `
          opacity: .5;
      `
    }
  }}
  input {
    display: none;
    & + ${SwitchToggle} {
      transform: translate(0);
    }

    &:checked + ${SwitchToggle} {
      transform: translate(calc(50% + 10px));
    }

    &:checked ~ ${SwitchBg} {
      background: ${({ theme, color }) => theme.colors[color]};
      opacity: 1;
      transform: translate(0);
    }
  }
`

const Switch = (props) => {
  const { onChange, checked, id, loading = false, color = 'primary', ...rest } = props
  return (
    <Wrapper htmlFor={id} loading={loading} color={color} {...rest}>
      <input type="checkbox" id={id} checked={checked} onChange={onChange} {...rest} />
      <SwitchToggle color={color} loading={loading} />
      <SwitchBg color={color} />
    </Wrapper>
  )
}

export default Switch
