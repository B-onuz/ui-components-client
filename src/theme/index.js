const colors = {
  red: '#db2828',
  lightRed: '#ef4d5a',
  mediumRed: '#FF4F5D',
  darkRed: '#9c2543',
  fullDarkRed: '#550e33',
  orange: '#f2711c',
  yellow: '#fbbd08',
  olive: '#b5cc18',
  green: '#21BA45',
  teal: '#00b5ad',
  blue: '#2185d0',
  facebookBlue: '#3B5998',
  blueLink: '#4285F4',
  violet: '#6435c9',
  purple: '#a333c8',
  pink: '#e03997',
  brown: '#a5673f',
  grey: '#767676',
  textGrey: '#8D8D8D',
  lightGrey: '#6D6D6D',
  mediumGrey: '#838383',
  darkGrey: '#9A9A9A',
  boldGray: '#55555A',
  black: '#1b1c1d',
  white: '#fdfdfd',
  warning: '#7A4D05',
  success: '#1E561F',
  info: '#0E566C',
  danger: '#973937',
  cyan: '#1CB5AD',
  disabledGrey: '#ffffff1f',
  disabledGreyText: '#ffffff4d',
}

colors.primary = colors.lightRed
colors.secondary = colors.fullDarkRed
colors.default = '#e0e1e2'

const breakpoints = ['40em', '52em', '64em', '80em']

breakpoints.sm = breakpoints[0]
breakpoints.md = breakpoints[1]
breakpoints.lg = breakpoints[2]
breakpoints.xl = breakpoints[3]

const fontSizes = [12, 14, 16, 20, 24, 32]

export default {
  fontFamily: `'Lato', sans-serif`,
  colors,
  fontSizes,
  breakpoints,
}
