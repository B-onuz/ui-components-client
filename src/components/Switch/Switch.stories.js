import React from 'react'
import Switch from './Switch'

export default {
  title: 'Switch',
  component: Switch,
}

export const SwitchExample = () => (
  <>
    <Switch m={10} color={'primary'} />
    <Switch m={10} color={'secondary'} />
    <Switch m={10} color={'olive'} />
    <Switch m={10} color={'blue'} />
    <Switch m={10} color={'green'} />
    <Switch m={10} color={'yellow'} />
    <Switch m={10} color={'danger'} />
  </>
)
