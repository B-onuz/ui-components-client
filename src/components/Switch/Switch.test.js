import React, { useState } from 'react'
import Switch from './Switch.js'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import theme from '../../theme'
import 'jest-styled-components'

test('Should render Switch when checked', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Switch checked={true} onChange={(e) => onChange(e.target.checked)} />
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Should render Switch when unchecked', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Switch checked={false} onChange={(e) => onChange(e.target.checked)} />
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
