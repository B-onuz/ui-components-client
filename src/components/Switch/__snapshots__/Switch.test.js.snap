// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Switch when checked 1`] = `
.c2 {
  display: block;
  background: #fff;
  position: relative;
  border-radius: 100%;
  box-shadow: 0 1px 3px rgba(0,0,0,0.4);
  height: 20px;
  width: 20px;
  top: 0;
  display: block;
  -webkit-transform: translate(25px);
  -ms-transform: translate(25px);
  transform: translate(25px);
  -webkit-transition: 0.2s;
  transition: 0.2s;
  z-index: 2;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  -webkit-transform: translate(calc(-55% - 25px));
  -ms-transform: translate(calc(-55% - 25px));
  transform: translate(calc(-55% - 25px));
  -webkit-transition: 0.25s ease-out;
  transition: 0.25s ease-out;
  opacity: 0;
}

.c0 {
  border-color: rgb(223,223,223);
  box-shadow: rgb(223,223,223) 1px 1px 3px 0px inset;
  -webkit-transition: border 0.4s,box-shadow 0.4s;
  transition: border 0.4s,box-shadow 0.4s;
  background-color: #f7f7f7;
  border: 1px solid #dfdfdf;
  border-radius: 20px;
  cursor: pointer;
  display: inline-block;
  height: 20px;
  position: relative;
  vertical-align: middle;
  width: 40px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  box-sizing: content-box;
  background-clip: content-box;
  -webkit-mask-image: -webkit-radial-gradient(white,black);
}

.c0 input {
  display: none;
}

.c0 input + .c1 {
  -webkit-transform: translate(0);
  -ms-transform: translate(0);
  transform: translate(0);
}

.c0 input:checked + .c1 {
  -webkit-transform: translate(calc(50% + 10px));
  -ms-transform: translate(calc(50% + 10px));
  transform: translate(calc(50% + 10px));
}

.c0 input:checked ~ .c3 {
  background: #ef4d5a;
  opacity: 1;
  -webkit-transform: translate(0);
  -ms-transform: translate(0);
  transform: translate(0);
}

<label
  className="c0"
  color="primary"
  loading={false}
>
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    className="c1 c2"
    color="primary"
    loading={false}
  />
  <div
    className="c3 c4"
    color="primary"
  />
</label>
`;

exports[`Should render Switch when unchecked 1`] = `
.c2 {
  display: block;
  background: #fff;
  position: relative;
  border-radius: 100%;
  box-shadow: 0 1px 3px rgba(0,0,0,0.4);
  height: 20px;
  width: 20px;
  top: 0;
  display: block;
  -webkit-transform: translate(25px);
  -ms-transform: translate(25px);
  transform: translate(25px);
  -webkit-transition: 0.2s;
  transition: 0.2s;
  z-index: 2;
}

.c4 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  -webkit-transform: translate(calc(-55% - 25px));
  -ms-transform: translate(calc(-55% - 25px));
  transform: translate(calc(-55% - 25px));
  -webkit-transition: 0.25s ease-out;
  transition: 0.25s ease-out;
  opacity: 0;
}

.c0 {
  border-color: rgb(223,223,223);
  box-shadow: rgb(223,223,223) 1px 1px 3px 0px inset;
  -webkit-transition: border 0.4s,box-shadow 0.4s;
  transition: border 0.4s,box-shadow 0.4s;
  background-color: #f7f7f7;
  border: 1px solid #dfdfdf;
  border-radius: 20px;
  cursor: pointer;
  display: inline-block;
  height: 20px;
  position: relative;
  vertical-align: middle;
  width: 40px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  box-sizing: content-box;
  background-clip: content-box;
  -webkit-mask-image: -webkit-radial-gradient(white,black);
}

.c0 input {
  display: none;
}

.c0 input + .c1 {
  -webkit-transform: translate(0);
  -ms-transform: translate(0);
  transform: translate(0);
}

.c0 input:checked + .c1 {
  -webkit-transform: translate(calc(50% + 10px));
  -ms-transform: translate(calc(50% + 10px));
  transform: translate(calc(50% + 10px));
}

.c0 input:checked ~ .c3 {
  background: #ef4d5a;
  opacity: 1;
  -webkit-transform: translate(0);
  -ms-transform: translate(0);
  transform: translate(0);
}

<label
  className="c0"
  color="primary"
  loading={false}
>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    className="c1 c2"
    color="primary"
    loading={false}
  />
  <div
    className="c3 c4"
    color="primary"
  />
</label>
`;
